src/coreclr/src/vm/weakreferencenative.cpp

```diff
@@ -132,21 +133,35 @@ IWeakReference* GetWinRTWeakReference(OBJECTREF* pObject)
 
     MethodTable* pMT = (*pObject)->GetMethodTable();
 
-    // If the object is not an RCW, then we do not want to use a WinRT weak reference to it
-    if (!pMT->IsComObjectType())
-    {
-        return nullptr;
-    }
+    SafeComHolder<IWeakReferenceSource> pWeakReferenceSource(nullptr);
 
-    // If the object is a managed type deriving from a COM type, then we also do not want to use a WinRT
+    // If the object is not an RCW, then we do not want to use a native COM weak reference to it
+    // If the object is a managed type deriving from a COM type, then we also do not want to use a native COM
     // weak reference to it.  (Otherwise, we'll wind up resolving IWeakReference-s back into the CLR
     // when we don't want to have reentrancy).
-    if (pMT != g_pBaseCOMObject && pMT->IsExtensibleRCW())
+    if (pMT->IsComObjectType()
+     && (pMT == g_pBaseCOMObject || !pMT->IsExtensibleRCW()))
     {
-        return nullptr;
+        pWeakReferenceSource = reinterpret_cast<IWeakReferenceSource*>(GetComIPFromObjectRef(pObject, IID_IWeakReferenceSource, false /* throwIfNoComIP */));
```

Comment:
Can this ever return non-null when the interopinfo on sync block is NULL? I am wondering whether we can have a short-circuit for null interopinfo up-front (potentially even in the caller of this method) and do this expensive COM related stuff only when the interop info is non-null.

Reply:
Or maybe there are even better ways how to make this pay-for-play by e.g. having the types that should participate in this marked with a managed interface.